server/src/main/java/org/opensearch/action/search/AbstractSearchAsyncAction.java

```diff
@@ -449,7 +450,8 @@ private void onShardFailure(final int shardIndex, @Nullable SearchShardTarget sh
         // we always add the shard failure for a specific shard instance
         // we do make sure to clean it on a successful response from a shard
         onShardFailure(shardIndex, shard, e);
-        final SearchShardTarget nextShard = shardIt.nextOrNull();
+        SearchShardTarget nextShard = new FailOpenRouting(e, clusterState).findNext(shardIt);
```

Comment:
Can we verify that if weights are set only then branch off to `FailureAwareRouting`

Reply:
Right now, that is not the case. In all cases, the code flow moves to `FailureAwareRouting`  to fetch next shard copy in case of failures. Should be fine I guess, do you think we should check for weights in action itself.